crates/axstorage/src/flex.rs

```diff
@@ -0,0 +1,376 @@
+    };
+    let callback = &mut *(user_data as *mut F);
+    callback(result);
+}
+
+/// Release storage.
+///
+/// This method should be called when done using the storage.
+///
+/// Note that the actual result of the release will be available in the callback.
+///
+/// # Parameters:
+/// - `storage`: [`Storage`] to release.
+/// - `callback`: Called when the release is done.
+pub fn release_async<F>(storage: &mut Storage, callback: Option<F>) -> Result<(), glib::Error>
+where
+    F: FnMut(Option<glib::Error>) + Send + 'static,
+{
+    // TODO: SAFETY
+    unsafe {
+        let callback = callback.map(|c| Box::into_raw(Box::new(c)));
+        let success = try_func!(
+            ax_storage_release_async,
+            storage.raw,
+            if callback.is_none() {
+                None
+            } else {
+                Some(release_async_trampoline::<F>)
+            },
+            match callback {
+                None => ptr::null_mut(),
+                Some(callback) => callback as gpointer,
+            }
+        );
+        debug_assert_eq!(success, GTRUE);
+        Ok(())
+    }
+}
+
+// TODO: SAFETY
+unsafe extern "C" fn release_async_trampoline<F>(user_data: gpointer, error: *mut GError)
+where
+    F: FnMut(Option<glib::Error>) + Send + 'static,
+{
+    let error = if error.is_null() {
+        None
+    } else {
+        Some(glib::Error::from_glib_full(error))
+    };
+    let callback = &mut *(user_data as *mut F);
+    callback(error);
+}
+
+/// Returns the location on the storage where the client should save its files.
+pub fn get_path(storage: &mut Storage) -> Result<CString, glib::Error> {
+    // TODO: SAFETY
+    unsafe {
+        let path = try_func!(ax_storage_get_path, storage.raw);
+        Ok(CString::from_raw(path))
+    }
+}
+
+/// Returns the status of the provided event.
+pub fn get_status(storage_id: &mut GStringPtr, event: StatusEventId) -> Result<bool, glib::Error> {
+    // TODO: SAFETY
+    unsafe {
+        let status = try_func!(
+            ax_storage_get_status,
+            to_mut_ptr(storage_id),
+            event.into_raw()
+        );
+        Ok(status == GTRUE)
+    }
+}
+
+/// Returns the storage ID.
+pub fn get_storage_id(storage: &mut Storage) -> Result<GStringPtr, glib::Error> {
+    // TODO: SAFETY
+    unsafe {
+        let mut storage_id = try_func!(ax_storage_get_storage_id, storage.raw);
+        Ok(ptr::read(
+            &mut storage_id as *mut *mut gchar as *mut GStringPtr,
+        ))
+    }
+}
+
+/// Returns the storage type.
+pub fn get_type(storage: &mut Storage) -> Result<Type, glib::Error> {
+    // TODO: SAFETY
+    unsafe {
+        let storage_type = try_func!(ax_storage_get_type, storage.raw);
+        debug_assert_ne!(storage_type, AXStorageType_UNKNOWN_TYPE);
+        Ok(Type::from_raw(storage_type))
+    }
+}
+
+// TODO: Verify safety of passing the result of this to C
+fn to_mut_ptr(s: &mut GStringPtr) -> *mut gchar {
```

Comment:
I think it's worth noting that this assumes that GStringPtr is implemented as a transparent wrapper around gchar and will have unintended results if that is not the case.